src/iosrv/init.c

```diff
@@ -225,17 +234,83 @@ dss_tgt_nr_get(int ncores, int nr)
 static int
 dss_topo_init()
 {
+	uint depth;
+	int num_obj;
+	int num_cores_visited;
+	char *cpuset;
+	int k;
+	hwloc_obj_t corenode;
+
 	hwloc_topology_init(&dss_topo);
 	hwloc_topology_load(dss_topo);
 
 	dss_core_depth = hwloc_get_type_depth(dss_topo, HWLOC_OBJ_CORE);
 	dss_core_nr = hwloc_get_nbobjs_by_type(dss_topo, HWLOC_OBJ_CORE);
-	dss_tgt_nr = dss_tgt_nr_get(dss_core_nr, nr_threads);
+	depth = hwloc_get_type_depth(dss_topo, HWLOC_OBJ_NUMANODE);
+	num_obj = hwloc_get_nbobjs_by_depth(dss_topo, depth);
+
+	/* if we are not NUMA aware, fall back to non-NUMA implementation */
+	if (num_obj <= 0) {
+		D_PRINT("NUMA information unavailable.\n");
+		dss_tgt_nr = dss_tgt_nr_get(dss_core_nr, nr_threads);
+
+		if (dss_core_offset < 0 || dss_core_offset >= dss_core_nr) {
+			D_ERROR("invalid dss_core_offset %d "
+				"(set by \"-f\" option),"
+				" should within range [0, %d]",
+				dss_core_offset, dss_core_nr - 1);
+			return -DER_INVAL;
+		}
+		return 0;
+	}
+
+	if (numa_node > num_obj) {
+		D_ERROR("Invalid NUMA node selected. "
+			"Must be no larger than %d\n",
+			num_obj);
+		return -DER_INVAL;
+	}
+
+	numa_obj = hwloc_get_obj_by_depth(dss_topo, depth, numa_node);
+	if (numa_obj == NULL) {
+		D_ERROR("NUMA node %d was not found in the topology",
+			numa_node);
+		return -DER_INVAL;
+	}
 
-	if (dss_core_offset < 0 || dss_core_offset >= dss_core_nr) {
+	/* create an empty bitmap, then set each bit as we */
+	/* find a core that matches */
+	core_allocation_bitmap = hwloc_bitmap_alloc();
+	if (core_allocation_bitmap == NULL) {
+		D_ERROR("Unable to allocate core allocation bitmap\n");
+		return -DER_INVAL;
+	}
+
+	dss_num_cores_numa_node = 0;
+	num_cores_visited = 0;
+
+	for (k = 0; k < dss_core_nr; k++) {
+		corenode = hwloc_get_obj_by_depth(dss_topo, dss_core_depth, k);
+		if (corenode == NULL)
+			continue;
+		if (hwloc_bitmap_isincluded(corenode->allowed_cpuset,
+			numa_obj->allowed_cpuset) != 0) {
```

Comment:
Align with opening paren of the function

Reply:
Fixed